### PR TITLE
feat(ecocredit): add state migrations

### PR DESCRIPTION
**Pull request by aleem1314**
## Description

Closes: #678 

* migrate credit classes and issuers to ORM v1
* create projects for existing classes
* migrate credit batches to ORM v1
* migrate balance & supply to ORM v1
* migrate credit type, project, batch sequence tables to ORM v1
* wire ecocredit migrations with `server.Manager`.

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: x/ecocredit/migrations/v3/core.go
Comment (ryanchristo) [inline x/ecocredit/migrations/v3/core.go:138]: Is it possible to iterate over credit batches rather than manually inserting project information?

- For `Name`, we could track the number of projects per credit class and automatically add the project name using the same format when creating projects (i.e. `C0101` is the first project in the `C01` credit class).
- For `ClassId` the batch provides the class id.
- For `Admin`, the batch provides the issuer.
- For `ProjectLocation`, the batch provides the project location.
- For `Metadata`, we could leave blank and maybe we could squeeze in another pull request adding the ability to update the metadata field #979

We would also need to skip adding a new project if all fields (except for the project name) match a project that has already been inserted. Does this sound possible?

File: x/ecocredit/migrations/v3/core.go
Comment (ryanchristo) [inline x/ecocredit/migrations/v3/core.go:205]: Is there something blocking this todo?

File: x/ecocredit/migrations/v3/core.go
Comment (ryanchristo) [inline x/ecocredit/migrations/v3/core.go:260]: I don't think v0.46 will make it into this release but good to have this here.

File: x/ecocredit/module/module.go
Comment (ryanchristo) [inline x/ecocredit/module/module.go:165]: I think we can actually leave this at `2`. We did not use the consensus version in the last upgrade.

https://github.com/regen-network/regen-ledger/blob/v3.0.0/app/stable_appconfig.go#L51-L52

**Review by ryanchristo — COMMENTED** [review]
This is looking great. I think we can avoid manually adding projects, which may not be necessary for mainnet (no credit batches exist on chain yet) but it would be nice to support for redwood.

File: x/ecocredit/migrations/v3/core.go
Comment (ryanchristo) [reply x/ecocredit/migrations/v3/core.go:205]: Oof... Yea, not sure how we should do this. We don't want this to be user controlled and we can't query the event, so we would have to manually update it (...and we would need to do redwood separately). Adding this to #980.

Comment (ryanchristo) [reply x/ecocredit/migrations/v3/core.go:205]: Oof... Yea, this will probably need to be done manually along with the curator address for baskets. We don't want this to be user controlled and we can't query the event. Added this item to #980.

Comment (aleem1314) [reply x/ecocredit/migrations/v3/core.go:205]: Yes, This needs to be done manually. Can we leave it empty for existing credit batches?

File: x/ecocredit/migrations/v3/core.go
Comment (aleem1314) [reply x/ecocredit/migrations/v3/core.go]: > Is it possible to iterate over credit batches rather than manually inserting project information?

Yes. We can do this.

File: types/module/server/manager.go
Comment (aleem1314) [inline types/module/server/manager.go:208]:
```suggestion
// RunMigrations performs state migrations for registered modules.
```

**Review by ryanchristo — COMMENTED** [review]
On our product call today, we discussed how we want to handle state migrations for projects. I still think this automatic migration is a potential approach, which will reduce the amount of manual operations required and support the migrations on redwood (which would require separate manual operations).

There was a concern brought up by @clevinson and @aaronc about the potential of multiple projects with the same location. Although I understand the concern, I don't see this as a major issue given that we are the only credit class creators (and credit issuers) and if we do issue any credits prior to the upgrade, we have the ability to provide distinct project locations for each credit batch.

That being said, we may need to revert the automatic migrations I requested for projects. @clevinson and @aaronc can you confirm? If so, my mistake and I apologize for the extra work.

Edit: For now, it sounds like we want to keep the automatic project migrations but we will also need to perform manual updates following the state migrations for each chain.

File: x/ecocredit/migrations/v3/core.go
Comment (ryanchristo) [reply x/ecocredit/migrations/v3/core.go:169]: This comment was from before the product call. Pending a response from @clevinson and @aaronc.

File: types/module/server/manager.go
Comment (aaronc) [inline types/module/server/manager.go:210]: This needs to be sorted to run deterministically... maybe at this point we should just use the SDK's `AppModule` to not risk these kinds of issues

**Review by aaronc — CHANGES_REQUESTED** [review]

File: types/module/server/manager.go
Comment (ryanchristo) [reply types/module/server/manager.go:210]: There's only one handler, which is for the ecocredit module. In that case, is this still an issue?

File: x/ecocredit/migrations/v3/core.go
Comment (technicallyty) [inline x/ecocredit/migrations/v3/core.go:23]: should we also include the params migration here? that would close #965. It should just take the existing params, convert them to v1 proto, and then set the default AskDenom param (https://github.com/regen-network/regen-ledger/pull/969).

File: types/module/server/manager.go
Comment (aaronc) [reply types/module/server/manager.go:210]: Best to just do it now and not have to worry

File: x/ecocredit/migrations/v3/core.go
Comment (aleem1314) [reply x/ecocredit/migrations/v3/core.go:23]: > should we also include the params migration here? that would close #965. It should just take the existing params, convert them to v1 proto, and then set the default AskDenom param (#969).

We don't have any breaking changes in params (It is just a version bump). For `allowed_ask_denoms` we can set them in the upgrade handler.

File: x/ecocredit/migrations/v3/core.go
Comment (aaronc) [reply x/ecocredit/migrations/v3/core.go:124]: in the SDK orm we can use `DeleteBy` to accomplish the same thing. For the old orm, I don't know

File: app/stable_appconfig.go
Comment (ryanchristo) [inline app/stable_appconfig.go:76]: I think we can hold off on implementing this here and point to #980 as a followup.

File: types/module/server/manager.go
Comment (ryanchristo) [inline types/module/server/manager.go:215]: Might be nice to add a comment here as to why we are sorting (i.e. to prevent non-determinism).

File: x/ecocredit/migrations/v3/core.go
Comment (ryanchristo) [reply x/ecocredit/migrations/v3/core.go:169]: After further discussing the @clevinson on the product call, we decided this would be the best approach to support both mainnet and redwood and then if we needed to perform any additional operations for mainnet (#980), we would handle those updates in the upgrade handler. Once this change is made and the related test is updated, this pull request has my approval.

File: x/ecocredit/migrations/v3/core.go
Comment (ryanchristo) [inline x/ecocredit/migrations/v3/core.go:191]: I think we can remove this todo or replace it with a comment about metadata being handled in the upgrade handler on a per chain basis.

File: x/ecocredit/migrations/v3/core.go
Comment (ryanchristo) [inline x/ecocredit/migrations/v3/core.go:206]: I think we can remove this todo or replace it with a comment about issuance date being handled in the upgrade handler on a per chain basis.

**Review by ryanchristo — COMMENTED** [review]
The updates look good to me. Just one more change requested from me in regards to how the existence of projects is determined. We need to check both the class id and project location when determining if a project already exists.

File: x/ecocredit/migrations/v3/core.go
Comment (aleem1314) [reply x/ecocredit/migrations/v3/core.go]: updated [3cd4bb1](https://github.com/regen-network/regen-ledger/pull/957/commits/3cd4bb11930724de96e955890ddf9eef0d5261d1#diff-f31c5489648094d493af1a8cfe6393dd8292a82f4e0902c5b981d3126bf36aeaR165)

**Review by ryanchristo — APPROVED** [review]
Looks good to me. Opened #993 as a followup.

**Comment by ryanchristo** [comment]
@aaronc the change you requested has been made. Is this something we can unblock now?

**Review by technicallyty — APPROVED** [review]
looks good to me, nice work! @aleem1314 👍🏻 

do we need an issue to test this manually and/or with integration test? looks like the test is calling the migration directly, so we'll need a way to ensure it's all wired up correctly 😄

**Comment by ryanchristo** [comment]
> do we need an issue to test this manually and/or with integration test? looks like the test is calling the migration directly, so we'll need a way to ensure it's all wired up correctly smile

We won't be able to test manually until we have the upgrade handlers implemented (#980). Also added a section to the ecoredit module readiness checklist for testing state migrations against a temporary test network (#715).

**Pull request merged**
